### PR TITLE
[python] handle arbitrary length feature names in Python-package

### PR DESCRIPTION
**Pull request by StrikerRUS**
Refer to https://github.com/microsoft/LightGBM/pull/4256#discussion_r628743652.

### CODE REVIEW

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -37,6 +39,8 @@ def test_basic(tmp_path):
         if i % 10 == 0:
             print(bst.eval_train(), bst.eval_valid())
 
+    assert train_data.get_feature_name() == feature_names
```
Comment (StrikerRUS) [inline tests/python_package_test/test_basic.py:42]: Current `master` fails here with
```
BufferError: Allocated feature name buffer size (255) wasinferior to the needed size (1001).
```

Comment (jameslamb) [reply tests/python_package_test/test_basic.py:42]: >  I overlooked bad string construction in #4143 

ah!! I see now

```python
required_string_vuffer_size.value
```

Sorry about that. I must have looked at that diff 10 times and never noticed that mistake (probably my mistake). I think we should consider adding `pylint` in CI. I see this error running `pylint python-package/ | grep vuffer`.

```text
python-package/lightgbm/basic.py:3462:124: E0602: Undefined variable 'required_string_vuffer_size' (undefined-variable)
python-package/build/lib/lightgbm/basic.py:3460:124: E0602: Undefined variable 'required_string_vuffer_size' (undefined-variable)
```

I mentioned that last year and then forgot about it :grimacing: https://github.com/microsoft/LightGBM/pull/3377#issuecomment-691286374. I should have written up an issue...I'll do that right now.

Comment (jameslamb) [reply tests/python_package_test/test_basic.py:42]: I just created https://github.com/microsoft/LightGBM/issues/4308

Comment (StrikerRUS) [reply tests/python_package_test/test_basic.py:42]: Wow! Very nice catch! I haven't noticed this too 🙁 .

 I was speaking about "no whitespace between words" issue.
```
BufferError: Allocated feature name buffer size (255) wasinferior to the needed size (1001).
                                                   ------^------
```

**Review by jameslamb — APPROVED** [review]
Looks great, thank you! The tests make sense to me, and I think these calls are correct. I got a bit confused about the difference between `actual_*`, `required_*`, and `reserved_*`, but I can't think of better names.

I compared these changes to the equivalent code in `c_api` and in `R-package/src/lightgbm_R.cpp`, and they all seem to match (except for `LGBM_BoosterGetFeatureNames()`, which is not used in the R package today).

File: tests/python_package_test/test_basic.py (hunk 1 above)
Comment (jameslamb) [reply tests/python_package_test/test_basic.py:42]: haha oh! Then two mistakes :laughing:

**Pull request merged**